docs/en/cookbook/dql-custom-walkers.rst

```diff
@@ -215,3 +216,77 @@ huge benefits with using vendor specific features. This would still
+
+.. code-block:: php
+
+    <?php
+    use Doctrine\DBAL\ArrayParameterType;
+    use Doctrine\DBAL\ParameterType;
+    use Doctrine\DBAL\Types\BooleanType;
+    use Doctrine\DBAL\Types\Exception\ValueNotConvertible;
+    use Doctrine\DBAL\Types\Type;
+    use Doctrine\ORM\Query\AST;
+    use Doctrine\ORM\Query\SqlOutputWalker;
+
+    class InterpolateParametersSQLOutputWalker extends SqlOutputWalker
+    {
+        /** {@inheritdoc} */
+        public function walkInputParameter(AST\InputParameter $inputParam): string
+        {
+            $parameter = $this->getQuery()->getParameter($inputParam->name);
+            if ($parameter === null) {
+                return '?';
+            }
+
+            $value = $parameter->getValue();
+            /** @var ParameterType|ArrayParameterType|int|string $typeName */
+            /** @see \Doctrine\ORM\Query\ParameterTypeInferer::inferType() */
+            $typeName = $parameter->getType();
+            $platform = $this->getConnection()->getDatabasePlatform();
+            $processParameterType = static fn(ParameterType $type) => static fn($value): string =>
+                (match ($type) { /** @see Type::getBindingType() */
+                    ParameterType::NULL => 'NULL',
+                    ParameterType::INTEGER => $value,
+                    ParameterType::BOOLEAN => (new BooleanType())->convertToDatabaseValue($value, $platform),
+                    ParameterType::STRING, ParameterType::ASCII => $platform->quoteStringLiteral($value),
+                    default => throw new ValueNotConvertible($value, $type->name)
+                });
+
+            if (is_string($typeName) && Type::hasType($typeName)) {
+                return Type::getType($typeName)->convertToDatabaseValue($value, $platform);
+            }
+            if ($typeName instanceof ParameterType) {
+                return $processParameterType($typeName)($value);
+            }
+            if ($typeName instanceof ArrayParameterType && is_array($value)) {
+                $type = ArrayParameterType::toElementParameterType($typeName);
+                return implode(', ', array_map($processParameterType($type), $value));
+            }
+
+            throw new ValueNotConvertible($value, $typeName);
+        }
+    }
+
+Then you may get raw SQL with:
+
+.. code-block:: php
+
+    <?php
+    $query
+        ->where('t.int IN (:ints)')->setParameter(':ints', [1, 2])
+        ->orWhere('t.string IN (?0)')->setParameter(0, ['3', '4'])
+        ->orWhere("t.bool = ?1")->setParameter('?1', true)
+        ->orWhere("t.string = :string")->setParameter(':string', 'ABC')
+        ->setHint(\Doctrine\ORM\Query::HINT_CUSTOM_OUTPUT_WALKER, InterpolateParametersSQLOutputWalker::class)
+        ->getSQL();
+
+The where clause of returned SQL should be like:
+
+.. code-block:: sql
+
+    WHERE t0_.int IN (1, 2)
+       OR t0_.string IN ('3', '4')
+       OR t0_.bool = 1
+       OR t0_.string = 'ABC'
```

Comment:
Ok, then what? You should probably state the use case for this explicitly.

Reply:
This example would help https://stackoverflow.com/questions/2095394/doctrine-how-to-print-out-the-real-sql-not-just-the-prepared-statement and https://stackoverflow.com/questions/4155288/how-to-write-union-in-doctrine-2-0/79647475#79647475